src/js/ProviderManager.js

```diff
@@ -0,0 +1,55 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals async, DirectoryLinksProvider, Links, PlacesProvider, gUserDatabase, gPinnedLinks, gBlockedLinks*/
+
+/**
+ * Singleton that handles link providers.
+ */
+
+ "use strict";
+
+const ProviderManager = {
+
+  _initialized: false,
+
+  get initialized() {
+    return this._initialized;
+  },
+
+  set initialized(initialized) {
+    this._initialized = initialized;
+  },
+
+  init: async(function*(requestURL) {
```

Comment:
On November 26, 2015 at 5:52:39 AM, Marina Samuel (notifications@github.com) wrote:
> > On second thought, initializeNewTabPage() looks pretty different  
> from whats in gh-pages vs. your pull request. Probably need to  
> decide how that should look first before moving this initialization  
> over.

I'll also note that the WebIDL work also significantly changes initialization. At least, it should simplify it significantly for various things (fingers crossed). For sure, gathering the "initial state" of things will need to be split up a bit as each part will now be able to ask for data directly from the browser and often get it back synchronously.  